### PR TITLE
Added downloadable report formats checkstyle format .xml, pipe delimited...

### PR DESCRIPTION
**Pull request by phpkidindia**
... .csv, binary format .tap

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -145,12 +145,63 @@ grunt.initConfig({
...
+});
+
+```
 
 ## Contributing
 In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [Grunt](http://gruntjs.com/).
 
 ## Release History
 
+- 2015-2-3   - v0.7.0: Added downloadable report formats checkstyle format .xml, pipe delimited .csv, binary format .tap
```
Comment (zacechola) [inline README.md]: Don't need to make changes to the README here. This section is used for versioned releases only.

File: tasks/bootlint.js [hunk 2]
```diff
@@ -17,17 +21,44 @@ module.exports = function(grunt) {
...
+      reportTAP: null
     });
 
     var totalErrCount = 0;
     var totalFileCount = 0;
+    var wcount = 0;
+    
+    
+    var reportXML = options.reportXML;
+    delete options.reportXML;
```
Comment (zacechola) [inline tasks/bootlint.js]: There's no reason to do this in JS. At least not here. You can't force the garbage collector to run.

File: tasks/bootlint.js [hunk 3]
```diff
@@ -17,17 +21,44 @@ module.exports = function(grunt) {
...
+
+    var reportCSV = options.reportCSV;
+    delete options.reportCSV;
+   
+    var reportTAP = options.reportTAP;
+    delete options.reportTAP;
+    
+      var output = '';
+    var xml ='<?xml version="1.0" encoding="utf-8"?>\r\n<checkstyle>\r\n';
```
Comment (zacechola) [inline tasks/bootlint.js]: Tabbing is a bit off.

File: tasks/bootlint.js [hunk 4]
```diff
@@ -37,49 +68,105 @@ module.exports = function(grunt) {
...
           var output = false;
+	
 
           if (lint.elements) {
+		
             lint.elements.each(function (_, element) {
               var loc = element.startLocation;
               grunt.log.warn(filepath + ":" + (loc.line + 1) + ":" + (loc.column + 1), lintId, lint.message);
+		var errline= loc.line + 1;
+		if(lint.id[0]=='E') { var severity = 'Error'; var tapsev = 'not ok';  }
```
Comment (zacechola) [inline tasks/bootlint.js]: Note the variable scoping here.

File: tasks/bootlint.js [hunk 5]
```diff
@@ -37,49 +68,105 @@ module.exports = function(grunt) {
...
-      if (totalErrCount > 0 && !options.showallerrors) {
+      if (totalErrCount > 0) {
         grunt.log.writeln().fail(totalErrCount + " lint error(s) found across " + totalFileCount + " file(s).");
         grunt.log.writeln().fail('For details, look up the lint problem IDs in the Bootlint wiki: https://github.com/twbs/bootlint/wiki');
-      } else if (totalErrCount > 0 && options.showallerrors) {
-        grunt.fail.warn(totalErrCount + " lint error(s) found across " + totalFileCount + " file(s). ");
       } else {
         grunt.log.ok(totalFileCount + ' file(s) lint free.');
       }
+      if (reportXML) {
```
Comment (zacechola) [inline tasks/bootlint.js]: Lots of repeated code here that could be consolidated.

**Comment by zacechola** [comment]
A couple things:

- Rebase your code.
- Add unit tests for your additions, please.

But more generally speaking: I'm not sure if this functionality must be in this grunt wrapper. Seems like the kind of functionality better suited for upstream in the [bootlint](http://github.com/twbs/bootlint) project.

I do see value in producing a file, though, but I don't think we should need to maintain multiple file formats. For example, some linting/validation grunt plugins produce a .json file. Or in the case of JSHint, users can choose to output the results to any file of their choosing rather than to stdout (which could then be further parsed for particular formatting.)

File: tasks/bootlint.js [hunk 6]
```diff
@@ -37,49 +68,105 @@ module.exports = function(grunt) {
...
+      }
+      if (reportCSV) {
+        grunt.util.hooker.unhook(process.stdout, 'write');
+        reportCSV = grunt.template.process(reportCSV);
+        var destination = path.dirname(reportCSV);
+        if (!grunt.file.exists(destination)) {
+          grunt.file.mkdir(destination);
+        }
+        grunt.file.write(reportCSV, csv);
+        grunt.log.ok('Report "' + reportCSV + '" created.');
```
Comment (cvrebert) [inline tasks/bootlint.js]: 'CSV Report "'

File: tasks/bootlint.js [hunk 7]
```diff
@@ -37,49 +68,105 @@ module.exports = function(grunt) {
...
           if (!output) {
             grunt.log.warn(filepath + ":", lintId, lint.message);
+                        lint.elements.each(function (_, element) {
+
+                          var loc = element.startLocation;
+
+            var errline= loc.line + 1
+		if(lint.id[0]=='E') { var severity = 'Error';  var tapsev = 'not ok'; }
+		if(lint.id[0]=='W') { var severity = 'Warning';  var tapsev = 'ok'; wcount++; }			
+		xml +='<error errcode="'+lint.id+'" message="'+lint.message+'" source="com.bootlint" />\r\n';
```
Comment (cvrebert) [inline tasks/bootlint.js]: com.getbootstrap.bootlint

File: tasks/bootlint.js [hunk 8]
```diff
@@ -37,49 +68,105 @@ module.exports = function(grunt) {
...
 
           if (lint.elements) {
+		
             lint.elements.each(function (_, element) {
               var loc = element.startLocation;
               grunt.log.warn(filepath + ":" + (loc.line + 1) + ":" + (loc.column + 1), lintId, lint.message);
+		var errline= loc.line + 1;
+		if(lint.id[0]=='E') { var severity = 'Error'; var tapsev = 'not ok';  }
+		if(lint.id[0]=='W') { var severity = 'Warning'; var tapsev = 'ok'; wcount++; }
+		xml +='<error line="'+ errline +'" errorcode="'+lint.id+'" message="'+lint.id +' '+lint.message+'" severity="'+severity+'" source="com.bootlint"/>\r\n';
```
Comment (cvrebert) [inline tasks/bootlint.js]: com.getbootstrap.bootlint

File: tasks/bootlint.js [hunk 9]
```diff
@@ -37,49 +68,105 @@ module.exports = function(grunt) {
...
+
+      if (reportTAP) {
+        grunt.util.hooker.unhook(process.stdout, 'write');
+        reportTAP = grunt.template.process(reportTAP);
+        var destination = path.dirname(reportTAP);
+        if (!grunt.file.exists(destination)) {
+          grunt.file.mkdir(destination);
+        }
+        grunt.file.write(reportTAP, tap);
+        grunt.log.ok('Tap Report "' + reportTAP + '" created.');
```
Comment (cvrebert) [inline tasks/bootlint.js]: Should be "TAP Report" (all-caps)?

File: README.md [hunk 10]
```diff
@@ -145,12 +145,63 @@ grunt.initConfig({
 });
 
 ```
+#### optioins.reportXML
```
Comment (cvrebert) [inline README.md]: typo: optioins => options

File: tasks/bootlint.js [hunk 11]
```diff
@@ -17,17 +21,44 @@ module.exports = function(grunt) {
...
+   
+    var reportTAP = options.reportTAP;
+    delete options.reportTAP;
+    
+      var output = '';
+    var xml ='<?xml version="1.0" encoding="utf-8"?>\r\n<checkstyle>\r\n';
+    var csv ='File|Line Number|Error Code|Severity\r\n';
+    var tap = "#Whole List Of BootLint Tests \r\n";
+    if (reportXML) {
+      grunt.util.hooker.hook(process.stdout, 'write', {
```
Comment (zacechola) [inline tasks/bootlint.js]: `grunt.util.hooker` is deprecated. See: http://gruntjs.com/api/grunt.util#grunt.util.hooker

**Comment by cvrebert** [comment]
@zacechola Perhaps this should be closed due to (a) lack of response/activity and (b) probably being out-of-scope?

**Comment by zacechola** [comment]
Yeah.

**Pull request closed without merging**